File: wrapper/android/src/main/java/com/zeticai/mlange/feature/AutoSelectPipelineFeature.kt
package com.zeticai.mlange.feature

abstract class AutoSelectPipelineFeature<T> : PipelineFeature<T> {
    val feature: PipelineFeature<T> by lazy {
        selectFeature()
    }

    protected abstract fun selectFeature(): PipelineFeature<T>

    override fun preprocess(input: Array<ByteArray>): Array<ByteArray> {
        return feature.preprocess(input)
    }

    override fun postprocess(output: Array<ByteArray>): T {
        return feature.postprocess(output)
    }

    override fun run(input: Array<ByteArray>): T {
        return feature.run(input)
    }
}

File: wrapper/ios/ext/Feature/AutoSelectPipelineFeature.swift
import Foundation

public class AutoSelectPipelineFeature<Input, Output>: PipelineFeature {
    lazy var feature: any PipelineFeature<Input, Output> = selectFeature()
    
    func selectFeature() -> any PipelineFeature<Input, Output> {
        fatalError("Must override selectFeature()")
    }
    
    public func preprocess(_ input: Input) -> [Data] {
        return feature.preprocess(input)
    }
    
    public func postprocess(_ output: [Data]) -> Output {
        return feature.postprocess(output)
    }
    
    public func run(_ input: Input) -> Output {
        return feature.run(input)
    }
}
